src/SSTFrequencyLimitDetector.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include "config.h"
+
+#include "SSTFrequencyLimitDetector.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+
+#include "geopm/Exception.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm_debug.hpp"
+
+namespace geopm
+{
+    enum Priorities {
+        HIGH_PRIORITY = 0,
+        MEDIUM_HIGH_PRIORITY = 1,
+        MEDIUM_LOW_PRIORITY = 2,
+        LOW_PRIORITY = 3,
+    };
+
+    SSTFrequencyLimitDetector::SSTFrequencyLimitDetector(
+        PlatformIO &platform_io,
+        const PlatformTopo &platform_topo)
+        : m_platform_io(platform_io)
+        , m_package_count(platform_topo.num_domain(GEOPM_DOMAIN_PACKAGE))
+        , m_core_count(platform_topo.num_domain(GEOPM_DOMAIN_CORE))
+        , m_clos_association_signals()
+        , m_sst_tf_enable_signals()
+        , m_cpu_frequency_max(m_platform_io.read_signal(
+              "CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0))
+        , m_cpu_frequency_sticker(m_platform_io.read_signal(
+              "CPU_FREQUENCY_STICKER", GEOPM_DOMAIN_BOARD, 0))
+        , m_cpu_frequency_step(m_platform_io.read_signal(
+              "CPU_FREQUENCY_STEP", GEOPM_DOMAIN_BOARD, 0))
+        , m_all_core_turbo_frequency(m_platform_io.read_signal(
+              "MSR::TURBO_RATIO_LIMIT:MAX_RATIO_LIMIT_7", GEOPM_DOMAIN_BOARD, 0))
```

Comment:
The `use_sst_tf_signals` thing is a best-effort approach to use SST-TF. It asks platformIO if the platform reports that it supports the feature so it can choose whether to use the SST detector instead of the TRL detector. "Unsupported" could mean "the IOGroup exists and reports that this feature is unsupported", or it could mean "this is so unsupported that the IOGroup doesn't even exist." The try-catch deals with that second case.

In either the SST or the TRL case, I still want exceptions from the detector itself to bubble up past `make_unique`. If the limit detector can't read these signals then it cannot make its limit-detection decisions, so there's no point finishing construction. Ultimately, construction of the balancer agent will fail due to this exception.

Reply:
Thanks for clarifying. Mainly wanted to make sure you were expecting these exceptions to propagate up.